assets/tests/TaskSchemaValidationService.js

```diff
@@ -0,0 +1,262 @@
+      verifyInputFunctionNameAppearsInAuxiliaryCodeIfNotNull: function(
+        task, auxiliaryCode) {
+        var inputFunctionName = task.getInputFunctionName();
+        if (inputFunctionName === null) {
+          return true;
+        }
+        var auxiliaryCodeClassName = 'AuxiliaryCode.';
+        if (inputFunctionName.startsWith(auxiliaryCodeClassName)) {
+          var functionName = inputFunctionName.substring(
+            auxiliaryCodeClassName.length);
+          return auxiliaryCode.indexOf(functionName) !== -1;
+        }
+        return false;
+      },
+      verifyOutputFunctionNameIsNullOrString: function(task) {
+        var outputFunctionName = task.getOutputFunctionName();
+        if (outputFunctionName === null) {
+          return true;
+        }
+        return (typeof outputFunctionName === 'string' ||
+          outputFunctionName instanceof String);
+      },
+      verifyOutputFunctionNameAppearsInAuxiliaryCodeIfNotNull: function(
+        task, auxiliaryCode) {
+        var outputFunctionName = task.getOutputFunctionName();
+        if (outputFunctionName === null) {
+          return true;
+        }
+        var auxiliaryCodeClassName = 'AuxiliaryCode.';
+        if (outputFunctionName.startsWith(auxiliaryCodeClassName)) {
+          var functionName = outputFunctionName.substring(
+            auxiliaryCodeClassName.length);
+          return auxiliaryCode.indexOf(functionName) !== -1;
+        }
+        return false;
+      },
+      verifyPrerequisiteSkillsAreArray: function(task) {
+        var prerequisiteSkills = task.getPrerequisiteSkills();
+        return Array.isArray(prerequisiteSkills);
+      },
+      verifyAcquiredSkillsAreArray: function(task) {
+        var acquiredSkills = task.getAcquiredSkills();
+        return Array.isArray(acquiredSkills);
+      },
+      verifyNoDuplicateSkillsInPrompts: function(task) {
+        var prerequisiteSkills = task.getPrerequisiteSkills();
+        var skillSet = new Set(prerequisiteSkills);
+        if (skillSet.size !== prerequisiteSkills.length) {
+          return false;
+        }
+        var acquiredSkills = task.getAcquiredSkills();
+        skillSet = new Set(acquiredSkills);
+        return skillSet.size === acquiredSkills.length;
+      },
+      verifyAtLeastOneAcquiredSkill: function(task) {
+        var acquiredSkills = task.getAcquiredSkills();
+        return acquiredSkills.length > 0;
+      },
+      verifyCorrectnessTestsAreArray: function(task) {
+        var correctnessTests = task.getCorrectnessTests();
+        return Array.isArray(correctnessTests);
+      },
+      verifyAtLeastOneCorrectnessTest: function(task) {
+        var correctnessTests = task.getCorrectnessTests();
+        return correctnessTests.length > 0;
+      },
+      verifyCorrectnessTestsHaveNonEmptyAllowedOutputArrays: function(task) {
+        var correctnessTests = task.getCorrectnessTests();
+        for (var i = 0; i < correctnessTests.length; i++) {
```

Comment:
Could be simplified? (Again, optional, but may be nice to have less code to read, going forward and for other tests.) Ditto below.

```
return correctnessTests.every(function(test) {
  return (
    angular.isArray(test.getAllAllowedOutputs()) &&
    test.getAllAlowedOutputs.length > 0); 
});
```